nfo_lookup plugin: Get metadata from nfo file to aid IMDB search

darcamo opened the pull request:
### Motivation for changes:

New Plugin: nfo_lookup

When searching for metadata with the imdb_lookup plugin and the entries coming
from the filesystem I had many wrong values, as well as many movies with no
match. The problem is that my files are named in portuguese and, although IMDB
lookup still works, it is less reliable since all it has is the
filename. Furthermore, sometimes there are different movies with the same name.

These movies are in my **Kodi library** and I already have
an ['nfo' file](http://kodi.wiki/view/NFO_files) for each of them containing,
among other things, the IMDB id of the correct movie. Since I couldn't find a
way to use this with flexget I decided to get my hands dirty.

### Detailed changes:

- Add a new 'nfo_lookup' plugin (in the metadata folder).
- Add a test file for it (`test_nfo_lookup.py`) including a folder with test
  data.
  
  
The idea is that this plugin can be used to add metadata from the nfo file to
the entry when the entries come from the filesystem input. It will also set the
'imdb_id' field of the entry such that if the imdb_lookup plugin is also used
it will be able to find the correct movie no matter how the movie file is
named. Currently it only works for movies (the nfo file must have a `<movie>`
tag) but it should be easy to extended to series.

The plugin will do the following:

- Get the name of the file associated with the entry
- Change the file extension to '.nfo', read the content of the nfo file 
- Add the metadata in the nfo file as 'nfo_something' fields in the entry
- Set the 'imdb_id' field of the entry

**Note**: This is working for me and I have implemented unittests for it, but
it should be tested by someone else with nfo files generated by a different
program than the one I'm using (MediaElch).

### Addressed issues:

None in particular besides my own needs. 

### Config usage if relevant (new plugin or updated schema):

```
nfo_lookup: yes
```

Note that the input must be the filesystem plugin. That is, the entry needs to
have a 'filename' and a 'location' fields.

Ex. of a full configuration

```
tasks:
  test_task:
    filesystem:
      path: some_path/
      mask: '*.mkv'
    nfo_lookup: yes
    imdb_lookup: yes
    <some_output_plugin>
```

### Log and/or tests output (preferably both):

```
paste output here
```

liiight commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,241 @@
+from flexget import plugin
+from flexget.event import event
+from flexget.plugins.metainfo.imdb_lookup import ImdbLookup
+# from flexget.utils.imdb import make_url as make_imdb_url
```
Unused imports

liiight commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,241 @@
+                           "thumb"]
+    # Fields that can appear multiple times in the nfo file
+    multiple_value_fields = [
+        # "thumb",
```
Probably need to be removed

darcamo replied:
Originally I was going to get other fields such as actors and directors as well from the nfo file. However, the nfo file only has the names and I would have to get the IMDB id of each of them to add to the fields `imdb_actors` and `imdb_directors`. 

For fields such as title and plot the nfo file might have different information, such as the translated title and plot. On the other hand, the names of actors and directors will be the same and thus I commented out this part to let the IMDB search fill the data.

If IMDB search is not performed anymore as part of this plugin then probably this should be kept and improved.

liiight commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,241 @@
+
+            # If there is no 'filename' field there is also no nfo file
+            if filename is None or location is None:
+                log.warning("Entry '{0}' didn't come from the filesystem plugin".format(entry.get('title')))
```
I think you could probably just return here, since there is no need to proceed if entry didn't originate from filesystem plugin.

darcamo replied:
The reason I'm not returning here is that I'm currently using this plugin as a
replacement to imdb_lookup and thus if I returned here there would be no
lookup.

cvium replied:
That's just bad design though

darcamo replied:
Maybe setting only `imdb_id` to help the imdb_lookup plugin and making sure that nfo_lookup executes before the imdb_lookup plugin would be a better option. 

Doing things as I did was a good first step for me to learn how things work and also use the make_html plugin, which works nicely with entries containing IMDB fields.

liiight commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,241 @@
+        # If there is already data from a previous parse then we don't need to
+        # do anything
+        if entry.get('nfo_id') is not None:
+            log.warning("Entry '{0}' was already parsed by nfo_lookup. "
```
Maybe warning isn't the Right level for this log?

liiight commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,241 @@
+        # ID we have already added the 'imdb_id' field and this will make the
+        # IMDB lookup below find the correct movie no matter what the title and
+        # filename are.
+        imdb_lookup = ImdbLookup()
```
You shouldn't use IMDb or any other hard coded lookups. Adding the fields to the entry should suffice. If the user wants to use any others lookup plugin he should so explicitly

darcamo replied:
The nice thing about doing like this is that I can use something like the make_html plugin and get a nice html output, where some data not in the nfo file come from the IMDB search (which will find the correct movie due to setting imdb_id before the search) and the data in the nfo file overwrites what was obtained from the search. This way I can have all the metadata with only the things I care from the nfo file, such as the title (now in my language) and the plot.

I agree that it is better to not do a hard lookup and use this plugin and the
imdb_plugin together to get everything. However, the nfo_lookup plugin needs to
act before imdb_lookup in order to set the `imdb_id` field (otherwise
imdb_lookup will find the wrong movie, since all it has is the filename). But,
when I was calling imdb_lookup it was overwriting all the imdb fields and thus
when I set the `imdb_title`, `imdb_plot_outline` and other fields in the
nfo_lookup plugin it was overwritten by the subsequent imdb_lookup. That's why
I decided to not using the imdb_lookup plugin and instead call it inside nfo_lookup.

Now I'm thinking that my mistake was trying to set the corresponding
`imdb_something` fields so that make_html works nicely. Maybe a better approach
is to set only `imdb_id` to help the imdb_lookup plugin (and maybe other
similar IDs such as for trakt) and only add fields corresponding to
`nfo_something`. Then If I want to use make_html I could create a different
jinja template that will use the corresponding `nfo_something` field if
available. If it is not available then fallback to `imdb_something`.

Is this reasonable?

liiight replied:
yeah, if `imdb_id` is populated it will be used for lookup by `imdb_lookup`. to make it run before lookups look into the `@priority` decorator

liiight commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,241 @@
+
+    @property
+    def movie_identifier(self):
+        """Returns the plugin main identifier type"""
```
This property method is not relevant for this plugin, since this is not a lookup identifier

darcamo replied:
I'm actually new to flexget and looking at the source I saw some plugins defining this property. It could be something important that I didn't see in the documentation and that's why I have included. I will remove this then.

liiight commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,241 @@
+
+@event('plugin.register')
+def register_plugin():
+    plugin.register(NfoLookup, 'nfo_lookup', api_ver=2, groups=['movie_metainfo'])
```
This should not be in the `movie_metainfo` group,

darcamo commented:
Overall I expected the code could have design problems, since I'm new to flexget and there are many details I still need to clarify.

For now I have removed the  movie_identifier property and the movie_metainfo group in the plugin registration.

Should I commit the "cassettes" created by the tests? I haven't used this before and I saw that there are many of them in the repository. On the other hand, If I refactor this to not call IMDB lookup anymore then I guess I would not have them anymore.

paranoidi commented:
@darcamo Some issues are to be expected, but don't sweat it. This is quite well made for first PR :)

paranoidi commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,234 @@
+
+    Use this only with nfo files you have created yourself.
+    """
+    # xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```
no separator comments in python code, empty lines where needed

paranoidi commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,234 @@
+                nfo_filename = self.get_nfo_filename(entry)
+                # Stop if there is no nfo file
+                if nfo_filename is None:
+                    log.warning(
```
logged usage should not use string formatting
```
log.warn('aaa %s', foo)
```

paranoidi commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,234 @@
+            # u'nfo_thumb': u'imdb_photo'
+        }
+
+        # multiple_field_mapping = {
```
remove leftovers

darcamo commented:
I think I have implemented the suggested changes.
Except for "generic to match any format of log file". The nfo file used by Kodi
has a know set of XML tags.

I have separated the code concerning reading the nfo file into a `NfoReader`
class, while the `NfoLookup` class is in charge of updating the entries. I will
improve `NfoReader` in the future to also handle series.

Right now there a dictionary "_fields" in `NfoReader` specifying which XML tags
should be read from the nfo file. One possible improvement to make the plugin
more generic is to allow specifying more tags that should be read and add them
to this dictionary.

For something other then XML files a different `SomethingReader` class could be
implemented, if desirable (but in that case wouldn't a different be better?).

I have also committed the cassettes and now all the tests are passing (there is
one test failing for python 2.7 in travis, but it is in a different plugin).

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+            # If there is no 'filename' field there is also no nfo file
+            if filename is None or location is None:
+                log.warning("Entry %s didn't come from the filesystem plugin",
+                            entry.get('title'))
```
Unnecessary line break here. We "allow" up to 120 characters.

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+                if nfo_filename is None:
+                    log.warning(
+                        "Entry %s has no corresponding %s file",
+                        entry.get('title'), self.nfo_file_extension)
```
Here as well

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+        # do anything
+        if entry.get('nfo_id') is not None:
+            log.warning("Entry %s was already parsed by nfo_lookup and it "
+                        "will be skipped. ", entry.get('title'))
```
Again

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+        if os.path.isfile(nfo_full_filename):
+            return nfo_full_filename
+        else:
+            return None
```
This is unnecessary. It will return None without the Else branch.

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+            self._nfo_filename = filename
+            self._root = root
+        else:
+            raise BadXmlFile()
```
Seems a bit silly to use the same exception for ParseError and if the file doesn't exist. At the very least you should give it a proper error message.

darcamo replied:
This exception is only used here when, for any reason, we can't read information from the filename that was specified. Therefore, NfoLookup can capture it and return without changing the entry. It's better then raising a general `Exception`, but I didn't think too much about it.

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+        if f is not None:
+            return getter_func(f)
+        else:
+            return None
```
Again, not needed

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+        if len(values) > 0:
+            return values
+        else:
+            return None
```
.

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+            multiple_bool = values[0]
+            getter_func = values[1]
+
+            nfo_field_name = u'nfo_{0}'.format(name)
```
I don't think the 'u' is necessary. Everything should already be unicode.

cvium commented on flexget/tests/test_nfo_lookup.py:
```diff
@@ -0,0 +1,328 @@
+#!/usr/bin/env python
```
Why is this here?

cvium commented on flexget/tests/test_nfo_lookup.py:
```diff
@@ -0,0 +1,328 @@
+            # Get all 'nfo' keys in the entry. In this test the info file only
+            # has the 'id' field
+            nfo_keys = sorted([i for i in entry.keys() if i[:3] == 'nfo'])
+            assert nfo_keys == [u'nfo_id']
```
The 'u' should not be needed on any of the strings

cvium commented on flexget/tests/test_nfo_lookup.py:
```diff
@@ -0,0 +1,328 @@
+            # Check that the 'nfo_id' field is set to the correct movie
+            assert entry['nfo_id'] == 'tt2316801'
+            assert entry['nfo_title'] == 'A Bela e a Fera'
+            assert entry['nfo_plot'] == u"Um romance inesperado floresce depois que a filha mais nova de um mercador em dificuldades se oferece para uma misteriosa besta com a qual seu pai ficou endividado."
```
This line is too long

cvium commented on flexget/tests/test_nfo_lookup.py:
```diff
@@ -0,0 +1,328 @@
+            assert entry['nfo_director'] == ["Christophe Gans"]
+            assert entry['nfo_genre'] == ["Fantasia", "Romance"]
+            assert entry['nfo_originaltitle'] == u"La Belle et la Bête"
+            assert entry['nfo_plot'] == u"Um romance inesperado floresce depois que a filha mais nova de um mercador em dificuldades se oferece para uma misteriosa besta com a qual seu pai ficou endividado."
```
Too long

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,256 @@
+    listed in the `_fields` attribute.
+    """
+    def __init__(self, filename):
+        from xml.etree.ElementTree import ParseError
```
Why is this import not in global scope?

darcamo commented:
I have implemented the new suggestion.

Now that I understand more about the code I see that I could not set `imdb_id`
at all and use the manipulate plugin to set `imdb_id` from the `nfo_id` field
to allow `imdb_lookup` to find the correct match. I tested and it works, but I
kept the code setting `imdb_id` because I don't see a drawback and I think it
is easier for users like this (manipulate plugin is an advanced plugin, after
all).

cvium commented on flexget/tests/test_nfo_lookup.py:
```diff
@@ -297,7 +273,9 @@ def test_nfo_with_all_fields(self, execute_task):
             assert entry['nfo_genre'] == ["Fantasia", "Romance"]
             assert entry['nfo_originaltitle'] == u"La Belle et la Bête"
-            assert entry['nfo_plot'] == u"Um romance inesperado floresce depois que a filha mais nova de um mercador em dificuldades se oferece para uma misteriosa besta com a qual seu pai ficou endividado."
+            assert entry['nfo_plot'] == (u"Um romance inesperado floresce depois que a filha mais nova de um mercador "
```
Missed a couple of u's

darcamo replied:
Searched u' and replaced with '.
I forgot u".

Done.

cvium commented:
I don't see anything wrong with setting imdb_id if that's what you're parsing. Just make sure it's valid.

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -0,0 +1,226 @@
+        # If the IMDB id was found in the nfo file, set the imdb_id field of the entry. This will help the imdb_lookup
+        # plugin to get the correct data if it is also used.
+        if 'nfo_id' in fields:
+            entry.update({'imdb_id': fields['nfo_id']})
```
Maybe you should use some regex or whatever to verify that it is a valid imdb_id. There is probably a helper function somewhere in the imdb plugins.

darcamo replied:
There is `extract_id` in `utils.imdb.py`, but it matches "title IDs" (start with "tt") and "people IDs" (start with "nm"). Therefore, I added a `is_valid_imdb_title_id` method that only matches title IDs. The regex is simple: Just 'tt' followed by 7 digits.

cvium commented on flexget/plugins/metainfo/nfo_lookup.py:
```diff
@@ -104,6 +108,13 @@ def get_nfo_filename(self, entry):
             return nfo_full_filename
 
+    @staticmethod
+    def is_valid_imdb_title_id(value=None):
```
I think it makes more sense to put it in utils/imdb

darcamo replied:
I didn't want to touch code in other files, but that was were I first looked to see if such a function existed. I'll move the function to there and create a similar one for "person ID".

cvium commented on flexget/utils/imdb.py:
```diff
@@ -37,22 +37,26 @@ def is_imdb_url(url):
     Return True if `value` is a valid IMDB ID for a person.
     """
-    if not isinstance(value, basestring):
+    if value is None:
```
Why this check?

darcamo replied:
You are fast!
I was writing the message explaining the reason and after I finished I saw a new comment.

The reason is that we can write something as

```
if is_valid_imdb_title_id(entry.get('nfo_id')):
    do_some_work_using_nfo_id_field()
```

In case the entry does not have the field then `entry.get` returns `None`,
`is_valid_imdb_title_id` returns `False` and we don't try to use the field
value. If `is_valid_imdb_title_id` raises an exception for `None` then everyone
would have to test for that.

darcamo commented on flexget/utils/imdb.py:
```diff
@@ -37,6 +37,26 @@ def is_imdb_url(url):
+    Return True if `value` is a valid IMDB ID for titles (movies, series, etc).
+    """
+    if not isinstance(value, basestring):
+        return False
```
I'm now raising a `TypeError` in case `value` is not a string. However, I think it is better to not raise an exception and instead return `False` if `value` is `None`. The reason is that we can write something as

```
if is_valid_imdb_title_id(entry.get('nfo_id')):
    do_some_work_using_nfo_id_field()
```

In case the entry does not have the field then `entry.get` returns `None`,
`is_valid_imdb_title_id` returns `False` and we don't try to use the field
value. If `is_valid_imdb_title_id` raises an exception for `None` then everyone would have to test for that.

cvium replied:
Or you could just do `entry.get('nfo_id', '')` and avoid the None value altogether. It's not the responsibility of this function to hold the developer's hand.

darcamo replied:
Makes sense. I'll make the change.

liiight commented:
thanks for this! you wanna update the wiki?

darcamo commented:
Sure.
I'll have a look on it.

2016-10-19 7:06 GMT-03:00 Or Carmi <notifications@github.com>:

> thanks for this! you wanna update the wiki?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Flexget/Flexget/pull/1437#issuecomment-254770049>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AD4gqjHHnrWIWm-H5vSSAqVWzyUhjSDGks5q1eujgaJpZM4KSQZi>
> .
>
